### ISSUE TITLE
Add NuSpec tag to add Redist=false on references

### ISSUE THREAD

**Issue by AArnott** [NuGet/Home#452]
NuGet packages should be able to self-identify that the assemblies they carry should *not* be redistributed by the project/app. 

Consider a NuGet package that delivers assemblies that serve as an SDK for an application or some other runtime component, where that external target already has all the assemblies it needs. In fact the NuGet package's EULA may prohibit redistributing the assemblies acquired via the NuGet package since the only allowed target is that application that already has them. In this case, a `CopyLocal=false` might at first seem like the desired behavior to apply to these references, but in fact that can break the developer's test projects, where those assemblies *are* needed with `CopyLocal=true` since the unit tests may run outside the application. So if CopyLocal (or Private) must be true, but packaging build .targets must not include these assemblies, there must be another indicator.

I propose that nuspec files allow a `<Redist>false</Redist>` tag, which will cause the project that references assemblies from this package to add this as metadata to the reference (whether references that are burned into the project file or references resolved at build time). Then .targets that create an archive of the project outputs (such as VSIX, MSI, etc.) can omit them.

In particular, the VS SDK .targets will package up everything into a VSIX archive which the user will then upload to the VS Gallery so other users can install it. But these VSIXes should not (indeed: *must* not) include VS SDK assemblies in them, but it is perfectly legitimate and important to our shipping scenarios for CopyLocal=true for unit testing of these VS extensions.

**Comment by davidfowl** [comment]
This doesn't seem like something that should be expressed in nuget metadata. You can't possibly know ahead of time where the application will be deployed right? I understand the scenario, but it seems a bit too specific. 
I think being able to have a reference assembly in NuGet is fine. The underlying packaging mechanism (VS SDK targets) would have to know that the VS SDK assemblies isn't supposed to be packaged with the VSIX.

**Comment by AArnott** [comment]
The use case here isn't that the NuPkg consumer is creating an application. The nuget package is an SDK for extending an existing application. Therefore we know exactly where it will be deployed. While I have VS in mind, the idea is general to any application's SDK. 

In general if we want to go very broad, NuGet would probably do well to evolve into full SDK support, including carrying assemblies of varying architectures, compile-time vs. runtime, etc. But I think for now a small step of encoding whether a NuPkg comes with a redist license or not as a simple boolean property is sufficiently general and simple to satisfy our needs and NuGet's schedule.

**Comment by davidfowl** [comment]
Few questions:
- Is the SDK installed on the developer's box?
- The SDK needs to be installed to run anyways right? In the unit test scenario, does the NuGet package bring just the compile time reference or does it bring in more?

> In general if we want to go very broad, NuGet would probably do well to evolve into full SDK support, including carrying assemblies of varying architectures, compile-time vs. runtime, etc. But I think for now a small step of encoding whether a NuPkg comes with a redist license or not as a simple boolean property is sufficiently general and simple to satisfy our needs and NuGet's schedule.

Without full SDK support, what is the purpose of this NuGet package? Can you give an example of something that fits into this category?

**Comment by AArnott** [comment]
> Is the SDK installed on the developer's box?

The NuGet package _is_ the SDK.

> The SDK needs to be installed to run anyways right? In the unit test scenario, does the NuGet package bring just the compile time reference or does it bring in more?

No. The SDK assemblies just have to be in the bin directory of the built project for it to be enough to test with. Imagine these SDK assemblies have nothing but interop types and utilities defined. The user may have to mock up a few objects to implement the interfaces the SDK defines, but other than that they are free to unit test all they want by using the SDK assemblies in the unit test process.

> Without full SDK support, what is the purpose of this NuGet package? Can you give an example of something that fits into this category?

Yes: the VS SDK.

**Comment by davidfowl** [comment]
Sounds like its enough to compile and write unit tests but not run. I would be interested to see what other software falls into this category.

Who would look at this redist property? When is it consumed?

**Comment by AArnott** [comment]
> I would be interested to see what other software falls into this category.

That's a good point to ponder. I don't know how many apps have SDKs.

> Who would look at this redist property? When is it consumed?

Anyone who writes .targets that package up files from a project's output directory would be a candidate for honoring the redist=false metadata by omitting that file from the package. The VSIX packaging targets might be the first. `nuget.exe pack` itself might as well, since it can take an MSBuild project file as a parameter.

**Comment by yishaigalatzer** [comment]
I think in this case there could be a simpler solution, like including these references in a targets file, and add extra properties to them, then have the SDK project auto exclude them.

The point here, is that since there is no built in packaging API in visual studio, and the VSIX packaging is unique, it should probably handle it.

I'm also thinking that a nuget package doesn't necessarily have the advanced knowledge of the target host (perhaps in the narrow scenario of vsix it does) a very simple scenario is a package including a common dll like newtonsoft.json, that you cannot pack into a vsix.

**Comment by yishaigalatzer** [comment]
@AArnott I want to think about this again. The ask here as far as I understand is to have a do not distribute me further, so more of a publish metadata. 

How do we see this work across multiple build systems (like the .NET CLI and msbuild)?

**Comment by AArnott** [comment]
Sure, we could perhaps think of it as a publish metadata. These assemblies may (or may not) be required at runtime so deploying them is fine (on a local machine) with the app or test project. But a build or packaging step that conforms to a 'publish' semantic may honor this metadata that nuget could set on its references.

I can imagine how this would work with MSBuild (in both NuGet v2 and v3 world), but I am not familiar enough with .NET CLI to speculate on how that would work with this.

**Comment by yishaigalatzer** [comment]
Can we make an msbuild feature for now for the vsix builder? And as first step verify if it does what we want it to be. Then follow-up on the CLI scenarios with @ericstj and @davidfowl

**Comment by davidfowl** [comment]
We may not need this feature at all with the flags that we added to the nuspec and in `project.json` (not sure about packages.config). 

@AArnott see https://github.com/NuGet/Home/wiki/%5BSpec%5D-Managing-dependency-package-assets

/cc @emgarten

**Comment by AArnott** [comment]
I don't think so, @davidfowl. I'm familiar with this doc and it doesn't seem to help because what we want is for the package to self-identify rather than the user having to hand-edit the package reference later. From the doc you linked to:

> Include flags are defined on dependency edges, not on the packages themselves.

@yishaigalatzer: yes, I think working on the MSBuild integration with VSIX as the test bed and first adopter of it would make sense. I'm not on the VSIX owning team though, so I'll loop them in with you so we can see what capacity they have at the moment.

**Comment by yishaigalatzer** [comment]
@david this seems to be a similar problem to the shared framework, where you depend on stuff the target guarantees to have.

**Comment by AArnott** [comment]
(@yishaigalatzer I think you accidentally pinged the wrong david.)

**Comment by yishaigalatzer** [comment]
Yes I did :(

**Comment by davidfowl** [comment]
The behavior you want is quite strange though. You both want and don't want copy local. What we settled on for sdk type packages in the cli was a marker on the dependency edge that would basically be inferring some flags. The difference that behavior of not copying anything local is consistent for running unit tests and for publishing a package.

Even if this isn't the exact mechanism you need, whatever we come up with needs to fit. So it's possible that a package should be able to define what default the default is for it's reference edge. That is, how should it act when it exists as a dependency.

**Comment by AArnott** [comment]
Not quite. Sorry if it wasn't clear before. CopyLocal=true is always appropriate. This is about suppressing packaging. Today, most packaging systems in MSBuild take CopyLocal as their tip to package up an assembly. But what I'm asking for is another hint that, when present, would override that. That way, copylocal=true for these references as unit tests and other uses may require, but packaging like VSIX and MSI would always omit them.

**Comment by davidfowl** [comment]
So this has no effect on the nuget behavior at all. It's just a hint to build systems that may or may not have a publish feature?

**Comment by davidfowl** [comment]
Does this flag need to be manually applied to the main package and all dependencies or just the main package? What are the exact semantics?

**Comment by AArnott** [comment]
It probably has no effect on where _nuget_ places assemblies or the references it adds. 
But as nuget would have to be the software to recognize from a package's manifest (or lib/\* patterns perhaps) that certain assemblies being references should not be published, I would say that indeed nuget behavior _does_ change because it has to read that, and propagate it to MSBuild.

For all cases I know of now, a single flag in the .nuspec file (e.g. `<SuppressPublishingContents>true</>`) would be enough. However the flag is more accurately associated with the individual files rather than at the package level. And no, the flag should not propagate to other packages that it brings in. I may write a VS SDK that brings in a bunch of other dependencies, and I personally add two assemblies to my package. My own package's contents should not be published in the VSIX, but all the other dependencies certainly might. 

As a variation on that (that isn't specific to publishing): my package has a embeddable interop type assembly and runtime assemblies. I want to suppress CopyLocal _and_ publishing of the embeddable assembly since it's never needed at runtime.

**Comment by davidfowl** [comment]
> But as nuget would have to be the software to recognize from a package's manifest (or lib/\* patterns perhaps) that certain assemblies being references should not be published, I would say that indeed nuget behavior does change because it has to read that, and propagate it to MSBuild.

You mean pack not publish right? There's no such thing as a nuget publish. So far it just sounds like a bool that needs to be read from the nuspec and placed in the lock file.

> For all cases I know of now, a single flag in the .nuspec file (e.g. <SuppressPublishingContents>true</>) would be enough. However the flag is more accurately associated with the individual files rather than at the package level. And no, the flag should not propagate to other packages that it brings in. I may write a VS SDK that brings in a bunch of other dependencies, and I personally add two assemblies to my package. My own package's contents should not be published in the VSIX, but all the other dependencies certainly might.

So it's just behavior that it's extra build system metadata. The build system has to then decide what that metadata means to it. There are examples of this in the new contentFiles feature.

> As a variation on that (that isn't specific to publishing): my package has a embeddable interop type assembly and runtime assemblies. I want to suppress CopyLocal and publishing of the embeddable assembly since it's never needed at runtime.

That's a different thing altogether unfortunately. The way this would make sense to represent in the nuget package as a specific type of asset in both the package and lock file. Probably a link section. It's pretty specific and special.

**Comment by AArnott** [comment]
> You mean pack not publish right? 

No, I meant publish. But not the NuGet term (that doesn't exist, as you say). I'm talking about a build step that has a publish semantic. And yes, NuGet's role would be to read the bool flag and propagate that to MSBuild (via the lock file, presumably, yes).

> That's a different thing altogether unfortunately. The way this would make sense to represent in the nuget package as a specific type of asset in both the package and lock file. Probably a link section. It's pretty specific and special.

Yes, I agree. I hesitated to even bring it up here. Is there an issue tracking the /link: scenario yet? Or perhaps it's now supported since our last discussion several months ago on the topic.

**Comment by jainaashish** [comment]
@AArnott is this still valid with `PackageReference`? can it not be achieved via `PrivateAssets` now?

**Comment by AArnott** [comment]
It's still a problem, yes. The main point is that packages should *self-identify* this rather than the user having to manually edit the PackageReference to get the expected and required behavior.

**Comment by AArnott** [comment]
To be clear, I don't mind if the setting is encoded in the PackageReference, if that is what your design requires, but when using the UI to install the package, the nuspec file should influence what the default setting is to be encoded.

**Comment by jainaashish** [comment]
I understand your point and keeping it as p1 in backlog for now since we haven't had much feedback from community on this for last couple of years.

**Comment by AArnott** [comment]
Understandable. Most package authors aren't writing developmentOnly=true packages.

**Comment by jnm2** [comment]
My company has a few framework-like NuGet packages. They are SDK packages containing 309 assemblies to be referenced by plugins for an enterprise financial software package. You want the assemblies to be referenced for compiling and editor tooling, but you'd never want to distribute them any more than you'd distribute .NET Framework assemblies with your app. Copying them to build output makes it impossible to separate what should be deployed from what should not.

Right now I'm weakly working around the nuspec's lack of expressiveness with this target:

```xml
<Project>

  <Target Name="PreventCopyLocal-SomePackageId" AfterTargets="ResolveReferences">
    <ItemGroup>
      <ReferenceCopyLocalPaths Remove="@(ReferenceCopyLocalPaths)" Condition="'%(ReferenceCopyLocalPaths.NuGetPackageId)' == 'SomePackageId'" />
    </ItemGroup>
  </Target>

</Project>
```

The downside is that it fails when projects transitively reference the package through another project. I'd like transitive referencing to just work.

There's a final wrinkle: unit tests *do* need a few of the 309 assemblies to be copied locally, depending on what csc actually ended up referencing. Maybe you're on to something with the idea that the 309 assemblies are always copied to every project's build output, but we add the new concept of a NuGet-controlled publish and the 309 assemblies are never copied to the *publish* output.

#### Proposal 

What if we kept it simple though? The nuspec would contain `<referencesCopyLocal>false</referencesCopyLocal>` or equivalent and the build system would never allow anything in the lib folder to be copied to build output *unless* you opt in. Then only test projects would opt in, perhaps by directly rather than transitively referencing the package and overriding the RedistReferences value:

##### Foo.csproj – no references from this package are copied to build output

```xml
    <PackageReference Include="Some.SDK" Version="1.0.0" />
```

##### Foo.Tests.csproj – all references from this package are copied to build output

```xml
    <PackageReference Include="Some.SDK" Version="1.0.0" ReferencesCopyLocal="true" />
```

**Comment by zivkan** [comment]
I didn't read the whole thread carefully, but I believe PackageReference's `ref/` folder does what is being requested here. The assets are available at compile time only, and will not be copied to the build's output directory. I think what's missing is an easy way to pack a project and have the assembly go into the nupkg's ref folder instead of the lib folder. It is also not compatible with packages.config. Having said that, I haven't tested it. I'm not sure what NuGet does when it only finds a ref assembly, and no lib or runtimes assemblies. Maybe it's necessary to also have an empty `runtimes/any/any/lib/_._` file.

Anyway, removing the SDK area and adding pack, since it's not related to NuGet's SDK APIs.

**Comment by AArnott** [comment]
Thanks for reviewing, @zivkan. But this scenario is not satisfied simply by using `ref/`. Per my last sentence in the issue description:

> but it is perfectly legitimate and important to our shipping scenarios for CopyLocal=true for unit testing of these VS extensions.
